agent/xds/listeners_apigateway.go

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package xds
+
+import (
+	"fmt"
+	envoy_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_listener_v3 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_tls_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	"github.com/hashicorp/consul/agent/proxycfg"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/types"
+)
+
+func (s *ResourceGenerator) makeAPIGatewayListeners(address string, cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
+	var resources []proto.Message
+
+	readyUpstreamsList := getReadyUpstreams(cfgSnap)
+
+	for _, readyUpstreams := range readyUpstreamsList {
+		listenerCfg := readyUpstreams.listenerCfg
+		listenerKey := readyUpstreams.listenerKey
+		boundListener := readyUpstreams.boundListenerCfg
+
+		//TODO I'm positive this doesn't go here but I'm just trying to get this thing to work
+		if cfgSnap.APIGateway.ListenerCertificates == nil {
+			cfgSnap.APIGateway.ListenerCertificates = make(map[proxycfg.IngressListenerKey][]structs.InlineCertificateConfigEntry)
+		}
+
+		for _, certRef := range boundListener.Certificates {
+			cert, ok := cfgSnap.APIGateway.Certificates.Get(certRef)
+			if !ok {
+				continue
+			}
+			cfgSnap.APIGateway.ListenerCertificates[listenerKey] = append(cfgSnap.APIGateway.ListenerCertificates[listenerKey], *cert)
+		}
+
+		var isAPIGatewayWithTLS bool
+		var certs []structs.InlineCertificateConfigEntry
+		if cfgSnap.APIGateway.ListenerCertificates != nil {
+			certs = cfgSnap.APIGateway.ListenerCertificates[listenerKey]
+		}
+		if certs != nil {
+			isAPIGatewayWithTLS = true
+		}
```

Comment:
You can simplify these kinds of conditionals like this:

```suggestion
		var certs []structs.InlineCertificateConfigEntry
		if cfgSnap.APIGateway.ListenerCertificates != nil {
			certs = cfgSnap.APIGateway.ListenerCertificates[listenerKey]
		}
		isAPIGatewayWithTLS := certs != nil
```